Allow registering a blocking (external) job to the agent

ksaaskil opened the pull request:
- Adds `meeshkan` library methods `create_blocking_job(job_name=..., report_interval=...)` and function decorator `@meeshkan.as_blocking_job(job_name=..., report_interval_secs=...)`
- Put all new Proxy methods behind `proxy.external_jobs.method` to avoid cluttering the proxy API
- Usage:
```python
meeshkan_job = meeshkan.create_blocking_job(name="my job")

with meeshkan_job:
    # Perform calculations
    meeshkan.report_scalar("loss", loss)
    # Create image
    # meeshkan.report_image(image_path)  # I wish one could do this
```
or preferably
```python
@meeshkan.as_blocking_job(job_name=..., report_interval_secs=...)
def train():
     # Perform calculations
    meeshkan.report_scalar("loss", loss)
```

idantene commented:
See my [previous PR](https://github.com/Meeshkan/meeshkan-client/pull/131), maybe it's better to keep things in one spot.
In there, the basic notebook integration works, but the scheduling (and therefore also the notifications) aspect of things doesn't work. That's due to a pickling problem (when sending jobs using `Pyro4`).  
Would love some help there :+1:

ksaaskil commented:
Refactored and renamed stuff as `ExternalJob` to highlight this does not have to overlap with the notebook extension enabling scheduling and other fancy stuff. This hack enables doing 
```python
@meeshkan.as_job(name="test-job", report_interval=30)
def train():
    # Do stuff
    meeshkan.report_scalar("loss", loss)
```
and getting notifications for function start, function end, and for reported scalars. The experiment came from trying to use Meeshkan on a Kaggle kernel and noticing that the current tools are just a bit too invasive if one only needs notifications.

idantene commented on meeshkan/core/job/jobs.py:
```diff
@@ -38,6 +38,32 @@ def terminate(self):
+        self.description = desc
+
+    @staticmethod
+    def create(pid: int, name: str, poll_interval=BaseJob.DEF_POLLING_INTERVAL) -> 'ExternalJob':
```
Was just thinking we should probably remove the `create_job` from `Job` class as well and just have a factory. What do you think?

ksaaskil replied:
I have no problem with that, I'm just very used to having static methods in the class for creating instances and it feels like the most logical place for that. But surely something like JobStore would be the logical place for those factory methods, keeping track of all jobs. I would kinda also like to require the job name to be unique, just appending a suffix if the name clashes with a previous job. Would remove some ambiguity on if stuff like `meeshkan logs job-name` works as expected or not.

idantene replied:
Yeah that works (+1 for the naming clash).  
It could also possibly be embedded in the `BaseJob` class, if we implement some logic as to how to recognize the different job types. If we have a `JobStore` I would like for it have added functionality and not just creating `Job`s though.

ksaaskil replied:
Sure, I think it would naturally be responsible for all the job look-ups that scheduler and proxy API are doing at the moment like `find_job_id` and `__get_job_by_pid` etc.

idantene replied:
Good thinking. We could also implement the persistent `JobHistory` from the backlog in that.

ksaaskil commented on meeshkan/core/scheduler.py:
```diff
@@ -75,20 +73,24 @@ def wait_stop(self):
+                 event_loop: Optional[asyncio.AbstractEventLoop] = None):
         self._queue_processor = queue_processor
         self.submitted_jobs = dict()  # type: Dict[uuid.UUID, Job]
+        self.external_jobs = dict()  # type: Dict[uuid.UUID, ExternalJob]
```
This is a temporary hack until we have managed to get the monitoring and job look-up stuff out from `Scheduler`. It's not pretty but works for now, I think.

ksaaskil commented on meeshkan/core/api.py:
```diff
@@ -43,6 +72,15 @@ def __enter__(self):
+    def external_jobs(self):
+        return self._external_jobs
+
+    def register_with_pyro(self, daemon: Pyro4.Daemon, name: str):
```
Personally I think this kind of grouping of endpoints makes a lot of sense for the Pyro proxy to keep things clean. And if/when we add more "groups" like this, I think it's very clean to have their registration logic all under one method in the `Api` class so users registering `Api` only need to call one method and have everything taken care of, instead of having to remember to register all required instances. IMHO it's Api's personal business what it needs to register and what not. But of course I move the extra line to Service as well.

idantene replied:
A matter of semantics I guess. I tend to think the `Service` is responsible for daemonizing everything, and as such it's own responsiblities. Of course, if there are more items to register, it might make sense.
But to be honest, I'm not convinced registering the `ExternalJobs` is needed; we're not registering the `Scheduler` either, the `Api` class simply has access to it's internal objects as is. There's no communication over Pyro for the internals.

ksaaskil replied:
It won't work without the register, without it Pyro is trying to return the `ExternalJobsApi` object and not proxy for it.

idantene replied:
Hah, interesting. I guess it's serialization vs. Pyro registration.
Is it something we should then consider serializing and including in the `_build_api` function?
If not, should we consider simply registering all the other objects as well? :thinking:

ksaaskil commented:
I agree it would be ideal to have `Stop` work with ExternalJobs as well, we could probably achieve that with e.g. a [messagebus](https://github.com/irmen/Pyro4/tree/master/examples/messagebus) between the Python process and the agent. We could also fix the server soon so that `Stop` is not shown for jobs not supporting it yet (SageMaker, ExternalJob). And yes, I think we should change `BaseJob` to only inherit `Trackable` and have `Stoppable` being implemented only for jobs actually supporting it. It's a good point that clicking `Stop` for the external job will not have good endcome at the moment.

ksaaskil commented:
That's a good point, could perhaps also use a custom signal and add custom signal handler to avoid messing with anything else running in the process.

idantene commented:
It's a tricky business, since `IPython` already presents a level of indirection. The `SIGINT` just passes a `SIGINT` to the currently running cell, we just need to verify it's for the same notebook *kernel* (and not, say, the notebook *process*)
